src/software/embedded/motor_controller/stspin_motor_controller.cpp

```diff
@@ -0,0 +1,241 @@
+    {
+        LOG(WARNING) << "UNDER_VOLT: Software undervoltage";
+        motor_faults.insert(TbotsProto::MotorFault::UNDER_VOLT);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVER_TEMP))
+    {
+        LOG(WARNING) << "OVER_TEMP: Software over temperature";
+        motor_faults.insert(TbotsProto::MotorFault::OVER_TEMP);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::START_UP))
+    {
+        LOG(WARNING) << "START_UP: Start up failed";
+        motor_faults.insert(TbotsProto::MotorFault::START_UP);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SPEED_FDBK))
+    {
+        LOG(WARNING) << "SPEED_FDBK: Speed feedback fault";
+        motor_faults.insert(TbotsProto::MotorFault::SPEED_FDBK);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVER_CURR))
+    {
+        LOG(WARNING) << "OVER_CURR: Software overcurrent";
+        motor_faults.insert(TbotsProto::MotorFault::OVER_CURR);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SW_ERROR))
+    {
+        LOG(WARNING) << "SW_ERROR: Software error";
+        motor_faults.insert(TbotsProto::MotorFault::SW_ERROR);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SAMPLE_FAULT))
+    {
+        LOG(WARNING) << "SAMPLE_FAULT: Sample fault for testing purposes";
+        motor_faults.insert(TbotsProto::MotorFault::SAMPLE_FAULT);
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVERCURR_SW))
+    {
+        LOG(INFO) << "OVERCURR_SW: Software overcurrent";
+        motor_faults.insert(TbotsProto::MotorFault::OVERCURR_SW);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::DP_FAULT))
+    {
+        LOG(WARNING) << "DP_FAULT: Driver protection fault";
+        motor_faults.insert(TbotsProto::MotorFault::DP_FAULT);
+        drive_enabled = false;
+    }
+
+    return MotorFaultIndicator(drive_enabled, motor_faults);
+}
+
+double StSpinMotorController::readThenWriteVelocity(const MotorIndex& motor,
+                                                    const int& target_velocity)
+{
+    const int16_t speed = sendAndReceiveFrame(motor, StSpinOpcode::GET_SPEED, 0);
+
+    // SET_SPEEDRAMP expects the target motor speed to be in register ax.
+    sendAndReceiveFrame(motor, StSpinOpcode::MOV_AX,
+                        static_cast<int16_t>(target_velocity));
+
+    // SET_SPEEDRAMP expects the ramp time in millis to be in register bx.
+    // We do speed ramping ourselves in MotorService, so we just want to
+    // set the target speed without ramping (hence we set reg bx to 0).
+    sendAndReceiveFrame(motor, StSpinOpcode::MOV_BX, 0);
+
+    sendFrame(motor, StSpinOpcode::SET_SPEEDRAMP);
```

Comment:
it looks like we're doing 7 SPI transactions to read/write the speed for a single motor. SPI transactions are costly (because there's a kernel context switch and memory copies involved). We should be trying to do this in max 2 transactions. Suggest reworking the firmware API here